server/base/src/main/java/org/apache/accumulo/server/util/FindCompactionTmpFiles.java

```diff
@@ -0,0 +1,212 @@
+          Arrays.stream(files).forEach(fs -> matches.add(fs.getPath()));
+        } catch (IOException e) {
+          LOG.error("Error looking for tmp files in volume: {}", vol, e);
+        }
+        return null;
+      }));
+    }
+    svc.shutdown();
+
+    while (futures.size() > 0) {
+      UtilWaitThread.sleep(10_000);
+      Iterator<Future<Void>> iter = futures.iterator();
+      while (iter.hasNext()) {
+        Future<Void> future = iter.next();
+        if (future.isDone()) {
+          iter.remove();
+          try {
+            future.get();
+          } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException("Error getting list of tmp files", e);
+          }
+        }
+      }
+    }
+    svc.awaitTermination(10, TimeUnit.MINUTES);
+    LOG.trace("Found compaction tmp files: {}", matches);
+
+    // Remove paths of all active external compaction output files from the set of
+    // tmp files found on the filesystem. This must be done *after* gathering the
+    // matches on the filesystem.
+    for (DataLevel level : DataLevel.values()) {
+      context.getAmple().readTablets().forLevel(level).fetch(ColumnType.ECOMP).build()
+          .forEach(tm -> {
+            tm.getExternalCompactions().values()
+                .forEach(ecm -> matches.remove(ecm.getCompactTmpName().getPath()));
+          });
+    }
+    LOG.trace("Final set of compaction tmp files after removing active compactions: {}", matches);
+    return matches;
+  }
+
+  public static class DeleteStats {
+    public int success = 0;
+    public int failure = 0;
+    public int error = 0;
+  }
+
+  public static DeleteStats deleteTempFiles(ServerContext context, Set<Path> filesToDelete)
+      throws InterruptedException {
+
+    final ExecutorService delSvc = Executors.newFixedThreadPool(8);
+    final List<Future<Boolean>> futures = new ArrayList<>(filesToDelete.size());
+    final DeleteStats stats = new DeleteStats();
+
+    filesToDelete.forEach(p -> {
+      futures.add(delSvc.submit(() -> {
+        if (context.getVolumeManager().exists(p)) {
+          return context.getVolumeManager().delete(p);
+        }
+        return true;
+      }));
+    });
+    delSvc.shutdown();
+
+    int expectedResponses = filesToDelete.size();
+    while (expectedResponses > 0) {
+      UtilWaitThread.sleep(10_000);
+      Iterator<Future<Boolean>> iter = futures.iterator();
+      while (iter.hasNext()) {
+        Future<Boolean> future = iter.next();
+        if (future.isDone()) {
+          expectedResponses--;
+          iter.remove();
+          try {
+            if (future.get()) {
+              stats.success++;
+            } else {
+              stats.failure++;
+            }
+          } catch (ExecutionException e) {
+            stats.error++;
+            LOG.error("Error deleting a compaction tmp file", e);
+          }
+        }
+      }
+      LOG.debug("Waiting on {} responses", expectedResponses);
+    }
```

Comment:
Could remove this sleep and instead always call get.  This avoids uneeded sleeping when things are done.  Also adjusted the log message to be a bit more specific about what is being waited on.


```suggestion
      int expectedResponses = filesToDelete.size();
      Iterator<Future<Boolean>> iter = futures.iterator();
      while (iter.hasNext()) {
        Future<Boolean> future = iter.next();
          try {
            while(true){
              try{
                if (future.get(10, TimeUnit.SECONDS)) {
                  stats.success++;
                } else {
                  stats.failure++;
                }
                break;
              }catch(TimeoutException e) {
                 LOG.debug("Waiting on {} background delete operations", expectedResponses);
              }
            }
          } catch (ExecutionException e) {
            stats.error++;
            LOG.error("Error deleting a compaction tmp file", e);
          }
          expectedResponses--;
      }
```

Reply:
Could adjust the log stmt to count all that are not done and drop the expectedResponses variable.

```java
LOG.debug("Waiting on {} background delete operations", futures.stream().filter(f->!f.isDone()).count());
```

Reply:
> I didn't implement your suggestion, because my implementation checked each future without waiting, whereas yours would wait for the next one to complete before making progress.

The code will not progress until all futures are done. Was thinking calling get() is the most time efficient way to do this because it will wait the minimal time for any single future to be done.  If we wait on a future and other after it are done that is ok, because once get is called on those it will return immediately.  So calling only get should limit the wait time to the longest future and no more.

Reply:
The rationale behind my implementation is to reduce memory usage in the Manager as fast as possible. If you think that's not really a concern here, then I'm not opposed to implementing your suggestion.

Reply:
> The rationale behind my implementation is to reduce memory usage in the Manager as fast as possible. If you think that's not really a concern here, then I'm not opposed to implementing your suggestion.

I like that goal, its more important that minimizing the wait time.  Minimizing the wait time probably only benefits ITs.